web/src/peer_data.ts

```diff
@@ -448,3 +474,67 @@ export async function get_unique_subscriber_count_for_streams(
     }
     return valid_subscribers.size;
 }
+
+// This function might retry up to 5 times to fetch data, then will
+// give up. `num_attempts` should only be set in the recursive call.
+async function load_subscriptions_for_user_with_retry(
+    user_id: number,
+    num_attempts = 0,
+): Promise<void> {
+    const subscriptions_promise = (async () => {
+        await channel.get({
+            url: `/json/users/${user_id}/channels`,
+            success(raw_data) {
+                const subscriptions =
+                    fetch_user_subscriptions_response_schema.parse(raw_data).subscribed_channel_ids;
+                for (const stream_id of subscriptions) {
+                    add_subscriber(stream_id, user_id);
+                }
+                fetched_user_subscriptions.add(user_id);
+            },
+            error(xhr) {
+                if (xhr.status === 400) {
+                    blueslip.error("Bad request to fetch user's subscribed channels.", {
+                        user_id,
+                    });
+                    return;
+                }
+                pending_subscription_requests.delete(user_id);
+                if (num_attempts === 5) {
+                    blueslip.error("Failure fetching user's subscribed channels. Giving up.", {
+                        user_id,
+                    });
+                    return;
+                }
+                blueslip.warn("Failure fetching user's subscribed channels. Retrying.", {
+                    user_id,
+                });
+                num_attempts += 1;
+                const retry_delay_secs = util.get_retry_backoff_seconds(undefined, num_attempts);
+                setTimeout(() => {
+                    void load_subscriptions_for_user_with_retry(user_id, num_attempts);
+                }, retry_delay_secs * 1000);
```

Comment:
This retry logic does not seem hooked up to the promises correctly. As soon as you `return` below, doesn't that mean the `await`ing caller proceeds with its next step?